Tour mode choice

zneedell opened the pull request:
**EARLY ATTEMPT--NEEDS TO BE TESTED THOROUGHLY**

This PR introduces a `Tour Mode` category, which is distinct from a trip mode. The tour mode can either be `CAR_BASED`, `BIKE_BASED`, or `WALK_BASED`, and defines whether you are choosing to take a personal vehicle with you on the tour. Each Tour can have one Tour Mode in the same way that each Trip can have one Trip Mode. A relevant tour mode choice strategy is stored in the `BeamPlan`. 

If a person is on a tour with the tour mode already defined, the tour mode serves to limit the type of modes they can choose between (although in practice this doesn't really do anything, because e.g. if you already are on a `WALK_BASED` tour you shouldn't have access to any of your household vehicles until you get back home, and shared vehicle trips are still allowed on `WALK_BASED` tours). I've added the ability to assign a tour mode strategy an associated vehicle ID, which is intended to mimic the `tourPersonalVehicle` and potentially replace it eventually, but that isn't used yet.

The main change to simulation behavior happens when a person goes into `ChoosesMode` with no tour mode defined. Before getting routes from the router, the agent uses the skims to estimate the utility for every available trip on the upcoming tour and then choose which (if any) available household vehicle to use for the tour. This sets the TourMode for the tour and limits the available modes for all remaining trips.

One main goal of this is to eliminate the long walk trips that arise when we're using ActivitySim plans that e.g. look like `HOME -> short trip -> ACTIVITY 1 -> long trip -> ACTIVITY 2 ...`, where the agent chooses to walk or take transit for the first trip and is then stranded without a car for the second trip.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/lbnl-ucb-sti/beam/3536)
<!-- Reviewable:end -->

Xuan-1998 commented on src/main/scala/beam/router/Modes.scala:
```diff
@@ -245,3 +247,69 @@ object Modes {
+        extends BeamTourMode(
+          "car_based",
+          Car,
+          Seq[BeamMode](CAR, CAR_HOV2, CAR_HOV3),
```
How about drive_transit? This might result drive_transit be categorized into walk_based tour
![image](https://user-images.githubusercontent.com/58761221/172474284-50e033e6-894e-4fde-9251-dd7322f8c33a.png)

Xuan-1998 commented on src/main/scala/beam/router/Modes.scala:
```diff
@@ -245,3 +247,69 @@ object Modes {
+          Seq[BeamMode](CAR, CAR_HOV2, CAR_HOV3)
+        )
+
+    case object BIKE_BASED extends BeamTourMode("bike_based", Car, Seq[BeamMode](BIKE), Seq[BeamMode](BIKE))
```
Same as above, we don't have bike_transit?

zneedell replied:
Oh, wait a minute, this is definitely a typo and should be `BeamTourMode("bike_based", Bike, Seq[BeamMode](BIKE), Seq[BeamMode](BIKE))`. My bad!

zneedell replied:
Although it looks like maybe this has no effect

Xuan-1998 commented on src/main/scala/beam/router/Modes.scala:
```diff
@@ -245,3 +247,69 @@ object Modes {
+    override val values: immutable.IndexedSeq[BeamTourMode] = findValues
+
+    val enabledModes: Map[BeamMode, Seq[BeamMode]] =
+      Map[BeamMode, Seq[BeamMode]](CAR -> Seq(DRIVE_TRANSIT), BIKE -> Seq(BIKE_TRANSIT))
```
why no WALK here for availableFirstAndLastLegModes?

dimaopen committed:
Addressed peer review notes

haitamlaarabi committed:
fix

haitamlaarabi committed:
fmt

dimaopen committed:
Fixed int overflow

haitamlaarabi committed:
rollback change

JustinPihony committed:
More control of EC2 instances and budgeting (#3532)

* initial lambda code

* add instance blocker for non-used zones

* recent changes

* Other avenues of instance creation

* new slack notification

* Review fix and change to false default

* Sync against current lambda

* More syncing

alex-vv committed:
Move remaining utilities

alex-vv committed:
Add spreadsheetData lambda function

alex-vv committed:
Add data branch commit

alex-vv committed:
Fixes

alex-vv committed:
Merge branch 'develop' into alex-vv/#3449-save-git-commit

# Conflicts:
#	aws/build.gradle
#	aws/src/main/python/beam_lambda/lambda_function.py

alex-vv committed:
Restore file contents that were previously in LFS

alex-vv committed:
Deduct client_x509_cert_url

JustinPihony committed:
Update lambda to 3.x and sync (#3542)

* simulateBeam

* last bumps needed

* bump to latest python

haitamlaarabi committed:
update

alex-vv committed:
Make lambda compatible with existing deployments

Xuan-1998 committed:
moving beam.agentsim.agents.vehicles.BeamVehicle#isSharedVehicle to BeamVehicleType class

Xuan-1998 committed:
assign right type to private bikes

Xuan-1998 committed:
adjust peer review

Xuan-1998 committed:
fmt

Xuan-1998 committed:
fixing compilation error

alex-vv committed:
Merge branch 'develop' into alex-vv/#3449-save-git-commit

# Conflicts:
#	aws/src/main/python/beam_lambda/lambda_function.py

alex-vv committed:
move `NetworkRelaxationScenarioGenerator` to scripts

alex-vv committed:
Add explanation for `beam.physsim.network.removeIslands` setting

haitamlaarabi committed:
adding timeout

nikolayilyin committed:
budget override functionality from develop

nikolayilyin committed:
Update lambda_function.py

the change I made today in order to make helics working again - does not worth creating yet another PR for lambda changes

Xuan-1998 committed:
Merge pull request #3544 from LBNL-UCB-STI/Xuan/fixing-sharedbike-type

Xuan/fixing sharedbike type

nikolayilyin committed:
Merge pull request #3534 from LBNL-UCB-STI/inm/jupyter-utility-add

New utility notebooks added into jupyter folder

alex-vv committed:
Merge pull request #3537 from LBNL-UCB-STI/alex-vv/#3445-move-utilities

Move all utilities into an upper level folder for easier discovery

alex-vv committed:
Merge pull request #3531 from LBNL-UCB-STI/alex-vv/#3407-relaxation-network-scenario

Network relaxation scenario & spec

alex-vv committed:
Move commit detection into lambda

alex-vv committed:
Detect data commit from HEAD

JustinPihony committed:
More proactive and messaging tweaks to AWS (#3545)

* More proactive and messaging tweaks

* update beam dependencies

* Split up the lambdas

* Change the documentation

* syntax fix

* Fixes from actual event testing

* sync yaml

alex-vv committed:
Fix error in lambda

alex-vv committed:
Fixes

JustinPihony committed:
Force pushing simple lambda typo

Xuan-1998 committed:
initial

Xuan-1998 committed:
adding origin into populationAttributes

Xuan-1998 committed:
adding originActivity into multinomial logit

Xuan-1998 committed:
fixing PersonAgent.scala

Xuan-1998 committed:
deleting useless comments

Xuan-1998 committed:
fmt

Xuan-1998 committed:
delete unused parameter in PersonAgent

Xuan-1998 committed:
using collect instead of filter + map

Xuan-1998 committed:
delete merging commit

Xuan-1998 committed:
swaping destination and origination

Xuan-1998 committed:
fmt

Xuan-1998 committed:
fix bugs

Xuan-1998 committed:
using collect instead of filter + map again

Xuan-1998 committed:
fmt

nikolayilyin committed:
Update gradle.deploy.properties

safer defaults

Xuan-1998 committed:
simplify collect method

Xuan-1998 committed:
simplify from x to _

Xuan-1998 committed:
delete useless code

Xuan-1998 committed:
getting origin from origin and destination from destination

Xuan-1998 committed:
fmt

Xuan-1998 committed:
make collect easier in a right way

Xuan-1998 committed:
switching to faster and cleaner way

Xuan-1998 committed:
fix bugs

Xuan-1998 committed:
lifting trip Index

Xuan-1998 committed:
delete ununsed value

Xuan-1998 committed:
fmt

Xuan-1998 committed:
fix bugs

Xuan-1998 committed:
Merge pull request #3546 from LBNL-UCB-STI/Xuan/Bike-Hourly-Adjustment-Parameter-Based-on-new-develop

Xuan/bike hourly adjustment parameter based on new develop

alex-vv committed:
Merge pull request #3543 from LBNL-UCB-STI/alex-vv/#3449-save-git-commit

AWS: Save current git branch commit and git data branch commit into spreadsheet

alex-vv committed:
Add Jupyter AWS lambda

alex-vv committed:
Fix typo in deployment lambda

nikolayilyin committed:
Merge pull request #3548 from LBNL-UCB-STI/alex-vv/fix-lambda-typo

Fix typo in deployment lambda

nikolayilyin committed:
Merge pull request #3539 from LBNL-UCB-STI/do/#3538-kill-if-no-progress

Exiting Beam if no progress

nikolayilyin committed:
fix for permission issue with cloud-init log file

nikolayilyin committed:
Merge pull request #3549 from LBNL-UCB-STI/inm/cloud-init-log-permission-issue

Fix for permissions problem with cloud-init log file

JustinPihony committed:
force push aws tweak

nikolayilyin committed:
cpu and ram logging

nikolayilyin committed:
better timeout for cpu/ram log

nikolayilyin committed:
better timeout for cpu/ram log

nikolayilyin committed:
zipping the cpu/ram log

nikolayilyin committed:
more iterations

nikolayilyin committed:
+ link

alex-vv committed:
Add gradle task parameters

alex-vv committed:
Renamed function dir based on AWS lambda name

alex-vv committed:
Allow to run Jupyter together with BEAM via deploy task

nikolayilyin committed:
a script to display cpu and ram usage collected during simulation on AWS

nikolayilyin committed:
Merge pull request #3551 from LBNL-UCB-STI/inm/cpu_ram_logging_on_aws

CPU and RAM usage logging on aws

alex-vv committed:
Merge branch 'develop' into alex-vv/#3540-jupyter-aws

# Conflicts:
#	aws/src/main/python/beam_lambda/lambda_function.py

alex-vv committed:
Fixes

Xuan-1998 committed:
initial documentation for reposition

nikolayilyin committed:
fix for Elki dependency

alex-vv committed:
Run Jupyter as non root

alex-vv committed:
Run Jupyter as non root for `jupyterEC2`

heyrutvik committed:
updates viz module

nikolayilyin committed:
Merge pull request #3553 from LBNL-UCB-STI/inm/fix-for-elki-dependency

Dependency: fix for Elki dependency

dimaopen committed:
Storing state if awaitingResponse is not empty

dimaopen committed:
Merge remote-tracking branch 'origin/develop' into do/#3550-dump-scheduler-state

dimaopen committed:
Enabled exiting in case of beam got stuck by default; removed maxInitializationTimeBeforeConsideredStuckMin param.

heyrutvik committed:
fixes package names

alex-vv committed:
Update docs

nikolayilyin committed:
Merge pull request #3555 from LBNL-UCB-STI/do/#3550-dump-scheduler-state

Stroging the scheduler state in case of beam exists with an error

dimaopen committed:
Scheduler state files are moved to the output folder, stop Scheduler when forcibly shutdown happens

nikolayilyin committed:
Update build.gradle

default instance size which is big enough for most cases

nikolayilyin committed:
Merge pull request #3554 from LBNL-UCB-STI/rrp/visualization-module

updates sim visualization module

heyrutvik committed:
fix param name typo

heyrutvik committed:
adds production/seattle changes

heyrutvik committed:
fixes resource closing

dimaopen committed:
Scheduler saves thread dump.

heyrutvik committed:
updates other data repos

Xuan-1998 committed:
documentation for repositioning

nikolayilyin committed:
Merge pull request #3557 from LBNL-UCB-STI/rrp/3409-close-gtfs-mapdb

fixes resource closing

nikolayilyin committed:
Merge pull request #3556 from LBNL-UCB-STI/rrp/3529-fix-typo-multinomial

#3529 Fix param name typo

dimaopen committed:
beam lambda_function.py: added a cron job that stores a beam thread dump in case of issues.

heyrutvik committed:
fixes plan issue

alex-vv committed:
Use gradle local jupyter task to run jupyter in lambda

haitamlaarabi committed:
fixing multinomialLogit

haitamlaarabi committed:
fixing compilation error

haitamlaarabi committed:
Merge branch 'hl/gemini-develop-mnl' of github.com:LBNL-UCB-STI/beam into hl/gemini-develop-mnl

haitamlaarabi committed:
adding gemini infrstructure R and merge

haitamlaarabi committed:
fix more errors

dimaopen committed:
Removed outdated job from cron, renamed file that contains cron tasks

Xuan-1998 committed:
Merge remote-tracking branch 'origin/develop' into Xuan/Documentation-for-reposition

alex-vv committed:
Fix Jupyter start

nikolayilyin committed:
Merge pull request #3559 from LBNL-UCB-STI/do/#3550-external-beam-termination

External script at AWS EC2 instances that saves beam thread dump

haitamlaarabi committed:
fmt

heyrutvik committed:
runs fmt

Xuan-1998 committed:
Merge pull request #3560 from LBNL-UCB-STI/rrp/3482-fixes-invalid-plan-generation

Create separate plans based on planIndex and set "selected" only (valid) one of them

haitamlaarabi committed:
fixing bug and compilation error

Xuan-1998 committed:
Merge remote-tracking branch 'origin/develop' into Xuan/Documentation-for-reposition

haitamlaarabi committed:
logs

haitamlaarabi committed:
more logs

haitamlaarabi committed:
testing potential fix

haitamlaarabi committed:
removing logs

dimaopen committed:
Fix time passing over midnight

haitamlaarabi committed:
testing changes to scale up

haitamlaarabi committed:
fix compilation error

haitamlaarabi committed:
activating log

haitamlaarabi committed:
fmt

haitamlaarabi committed:
removing logs

nikolayilyin committed:
Merge pull request #3563 from LBNL-UCB-STI/do/fix-kill-beam-script

Fix for 'time goes over midnight' issue in simulateBeam lambda

alex-vv committed:
Merge branch 'develop' into alex-vv/#3540-jupyter-aws

# Conflicts:
#	aws/src/main/python/beam_lambda/lambda_function.py

alex-vv committed:
Permission fixes

alex-vv committed:
Merge pull request #3552 from LBNL-UCB-STI/alex-vv/#3540-jupyter-aws

Runs Jupyter at AWS, standalone or together with BEAM

frikit committed:
Fix for display long legend text for NonArrivedAgentsAtTheEndOfSimulation.png

haitamlaarabi committed:
fmt

haitamlaarabi committed:
updatin charging scale up

haitamlaarabi committed:
potential fix to bad completion

g-derzhavets committed:
[issue-3523] fix beam doesnt stop when exception occurs

g-derzhavets committed:
[issue-3523] scalafmt

nikolayilyin committed:
Merge pull request #3565 from LBNL-UCB-STI/frikit/fix-non-arrived-agents-legend-labels

Fix for display long legend text for NonArrivedAgentsAtTheEndOfSimula…

g-derzhavets committed:
[issue-3523] add cp thread dump to final path

g-derzhavets committed:
Merge branch 'develop' into gd/issue-3524_beam_doesnt_stop_working_when_runtime_exception

g-derzhavets committed:
[issue-3523] improve javadoc for try-catch in RunBeam

g-derzhavets committed:
[issue-3523] move sys.exit out of try-catch

g-derzhavets committed:
Merge pull request #3566 from LBNL-UCB-STI/gd/issue-3524_beam_doesnt_stop_working_when_runtime_exception

[issue-3523] fix beam doesn't stop when exception occurs

nikolayilyin committed:
a fix for terminateEC2, stopEC2 and startEC2 commands

alex-vv committed:
Refactor `jupyterEC2` task to use `simulateBeam` lambda

g-derzhavets committed:
[issue-3523] fix lambda copy thread dump

nikolayilyin committed:
Merge pull request #3568 from LBNL-UCB-STI/inm/aws_build_gradle_fix

AWS Gradle: a hotfix for terminateEC2, stopEC2 and startEC2 commands

Xuan-1998 committed:
Merge pull request #3558 from LBNL-UCB-STI/Xuan/Documentation-for-reposition

Xuan/documentation for reposition

frikit committed:
Upgrade gradle 7.1.1 to 7.5
upgrade sphinx 1.0.3.Final to 2.10.1
upgrade to latest beam utils 2.17 to 2.19
upgrade scala 2.12.10 to 2.12.16(beam-utils has the same version now)
Fix deprecated warnings from scala for new format method
remove deprecated jcenter per gradle docs
upgradedocker compose plugin

zneedell committed:
record correct (non-generalized) times and costs in asim skims

zneedell committed:
record correct (non-generalized) times and costs in asim skims

nikolayilyin committed:
Merge pull request #3570 from LBNL-UCB-STI/gd/fix_lambda

[issue-3523] fix lambda copy thread dump

nikolayilyin committed:
bigger timer

heyrutvik committed:
Merge pull request #3573 from LBNL-UCB-STI/inm/beam_stuck_guard_bigger_timer

Bigger timer for stuck guard - from 3 to 5 hours

alex-vv committed:
Fixes

alex-vv committed:
Merge branch 'develop' into alex-vv/rewrite-jupyter-ec2-task

alex-vv committed:
Fixes

alex-vv committed:
Fixes

zneedell committed:
assign each linkId to a TAZ using input shapefile, and use links to look up start/end TAZ for skims

nikolayilyin committed:
Update lambda_function.py

bigger timeout for stuck guard

nikolayilyin committed:
Update lambda_function.py

even bigger timeout

alex-vv committed:
Merge pull request #3571 from LBNL-UCB-STI/alex-vv/rewrite-jupyter-ec2-task

Refactor `jupyterEC2` task to use `simulateBeam` lambda

zneedell committed:
fmt

zneedell committed:
write missed linkIds

zneedell committed:
update logging for stuck guard

zneedell committed:
fix header

zneedell committed:
Throw skimmer events for non-chosen itineraries

zneedell committed:
fmt

zneedell committed:
Revert "update warmstart files for sfbay"

This reverts commit aaddf222 (failing on gcloud due to bad gzip)

haitamlaarabi committed:
update scripts

haitamlaarabi committed:
fmt

haitamlaarabi committed:
merging changes from mnl branch

haitamlaarabi committed:
fmt

haitamlaarabi committed:
merging develop

haitamlaarabi committed:
only CAVs can charge at depot

haitamlaarabi committed:
Merge remote-tracking branch 'origin/hl/gemini-develop-mnl' into gemini-develop

haitamlaarabi committed:
replan on the fly when no hh vehicles are available

zneedell committed:
fix tests

nikolayilyin committed:
Merge pull request #3577 from LBNL-UCB-STI/zn/skimmer-events-for-non-chosen-modes

Throw skimmer events for non-chosen itineraries

nikolayilyin committed:
Update gradle.deploy.properties

nikolayilyin committed:
Update gradle.deploy.properties

nikolayilyin committed:
Update gradle.properties

zneedell committed:
Merge pull request #3501 from LBNL-UCB-STI/zn/new-r5-options

allow non-dominating transit paths

zneedell committed:
add new config, add buffer to routing request

zneedell committed:
Merge branch 'develop' into zn/use-taz-geometries-for-skims

zneedell committed:
Merge pull request #3578 from LBNL-UCB-STI/zn/times-and-costs-in-asim-skimmer

Use [non-generalized] times and costs in asim skimmer

g-derzhavets committed:
[issue-3567] improve completion status in runs spreadsheet

frikit committed:
Merge pull request #3572 from LBNL-UCB-STI/frikit/gradle_upgrade

Upgrade gradle 7.1.1 to 7.5

nikolayilyin committed:
Merge remote-tracking branch 'origin/develop' into zn/use-taz-geometries-for-skims

nikolayilyin committed:
fix for a new method signature

zneedell committed:
handle missing lastUsedStall (new error for emergency vehicles for some reason)

zneedell committed:
include buffer in trip start time as well

haitamlaarabi committed:
addressing @nikolayilyin comments

haitamlaarabi committed:
Merge branch 'gemini-develop' of github.com:LBNL-UCB-STI/beam into gemini-develop

haitamlaarabi committed:
merge with develop

haitamlaarabi committed:
testing a fix

nikolayilyin committed:
a fix for a new constructor signature

g-derzhavets committed:
Merge branch 'develop' into gd/issue-3567_improve_completion_status_in_runs_spreadsheet

nikolayilyin committed:
Merge pull request #3575 from LBNL-UCB-STI/zn/use-taz-geometries-for-skims

Use TAZ geoms for skim origin/destination

zneedell committed:
bugfix -- only add buffer for transit trips

zneedell committed:
Merge remote-tracking branch 'origin/develop' into zn/drive-transit-buffer-time

zneedell committed:
fix mistake and fix tests

zneedell committed:
fix tests

zneedell committed:
fix tests

haitamlaarabi committed:
adding isCav to vehicleTypes

haitamlaarabi committed:
fix test

nikolayilyin committed:
Merge pull request #3580 from LBNL-UCB-STI/zn/drive-transit-buffer-time

Add buffer time for access to transit

g-derzhavets committed:
Merge pull request #3581 from LBNL-UCB-STI/gd/issue-3567_improve_completion_status_in_runs_spreadsheet

[issue-3567] improve completion status in runs spreadsheet

haitamlaarabi committed:
Merge pull request #3579 from LBNL-UCB-STI/gemini-develop

Latest code update from GEMINI-XFC work

nikolayilyin committed:
in case logger is not initialized yet

nikolayilyin committed:
in case logger is not initialized yet

nikolayilyin committed:
Merge pull request #3584 from LBNL-UCB-STI/inm/logg-exception-to-output-for-earlier-problems

Bugfix: in case logger is not initialized yet

alvinlee001 committed:
update config template file path in docs

nikolayilyin committed:
Merge pull request #3590 from LBNL-UCB-STI/alvin/update-dev-guide-docs-2022

update config template file path in docs

alex-vv committed:
Add parameters for gradle task to docs

nikolayilyin committed:
Merge pull request #3592 from LBNL-UCB-STI/alex-vv/3586-update-jupyter-docs

Add parameters for gradle task to docs

alvinlee001 committed:
added runtime validation (where household and plans are required in the old format) for legacy-beam-scenario-specific runs, while making these params optional in matsim.conf, so that we can reduce verbosity of config files

1) Scenario 1
- When: beam.conf or other legacy beam config not include household or plans
- Expected result: run fails, letting user know what config params is missing. Simulation did not proceed.

2) Scenario 2
- When: urbanism-v2/conf or other new beam config not include household or plans
- Expected result: run goes without error, simulation runs smoothly with expected result.

alvinlee001 committed:
fix formatter error

alvinlee001 committed:
test!

nikolayilyin committed:
Merge pull request #3593 from LBNL-UCB-STI/alvin/#3177-unnecessary-inputs-with-urbansim-scenarios

added runtime validation (where household and plans are required in t…

alex-vv committed:
Remove obsolete maven repositories

Also set mokito version to a fixed one to avoid excessive resolving.

nikolayilyin committed:
Merge pull request #3595 from LBNL-UCB-STI/alex-vv/fix-maven-repos

Remove obsolete maven repositories

alex-vv committed:
Fix skimTypes for skimmers

alex-vv committed:
Merge pull request #3597 from LBNL-UCB-STI/alex-vv/fix-skimmers-type

Fix skimTypes for skimmers

Xuan-1998 commented on src/main/scala/beam/router/Modes.scala:
```diff
@@ -245,3 +247,69 @@ object Modes {
+
+    // TODO: Also allow use of shared bikes/cars in walk based tours
+    case object WALK_BASED
+        extends BeamTourMode(
```
Think we could have 
`id: Id[BeamVehicleType]`
And then distinguish shared bike by
`def isSharedVehicle: Boolean = id.toString.startsWith("sharedVehicle")`
And then filter shared bike by
```
val bikeVehicleType = beamScenario.vehicleTypes.values
  .filterNot(_.isSharedVehicle)
  .find(_.vehicleCategory == VehicleCategory.Bike)
  .getOrElse(throw new RuntimeException("Bike not found in vehicle types."))
```
And then create a new object of shared bike to put in

zneedell replied:
I think you're on to something -- how about rather than giving this an `id` attribute we change the `allowedBeamModes` to a private attribute and give the class a new method like `def allowedBeamModes(vehicles: Vector[VehicleOrToken]): Seq[BeamMode]` that checks whether the vehicles are shared or not before returning the allowed beam modes, having `BIKE_BASED` not return `BIKE` if the only bikes available are shared bikes.

zneedell commented on src/main/scala/beam/agentsim/agents/modalbehaviors/ChoosesMode.scala:
```diff
@@ -259,6 +252,49 @@ trait ChoosesMode {
+            dummyVehicleType,
+            beamScenario.fuelTypePrices(dummyVehicleType.primaryFuelType)
+          )
+          val modeToTourMode =
```
@Xuan-1998 So as I understand it you're suggesting that the problem here is that if the agents have access to a shared bike when they hit this logic, then they end up potentially choosing a `BIKE_BASED` tour, even though they need to leave the bike behind after the first trip? That definitely seems like it could be causing the problem. Perhaps we need to add some other logic that only allows a `BIKE_BASED` tour if they have access to a personal bike (and similar for personal vs shared cars)

zneedell replied:
And just thinking through a to do here -- we don't currently factor in walking distance to the available street vehicles into these costs. Perhaps eventually we should call the skimmer to factor in those access/egress costs as well. Don't think it's worth it for now, though.

zneedell commented on src/main/scala/beam/agentsim/agents/modalbehaviors/ChoosesMode.scala:
```diff
@@ -259,6 +252,49 @@ trait ChoosesMode {
+        TourModeChoiceStrategy(chosenCurrentTourMode)
+      )
+
+      val availableModesGivenTourMode = getAvailableModesGivenTourMode(
```
I think the one other change here is that if the person still chooses a `BIKE_BASED` tour, we cannot let them use a shared bike. Perhaps we need to add another function that is `getAvailableVehiclesGivenTourMode` before this one is called that filters our shared vehicles if we're on a `BIKE_BASED` or `CAR_BASED` tour

Xuan-1998 commented:
Let's open this one to be ready to get slack notification

Xuan-1998 commented:
Let's open this one to be ready to get slack notification

Xuan-1998 committed:
typo correction

Xuan-1998 committed:
define the function to check vehicle shared or not

Xuan-1998 committed:
adding shared deny into shared bike

Xuan-1998 committed:
fix

Xuan-1998 committed:
checks whether the vehicles are shared or not before returning the allowed beam modes, having BIKE_BASED not return BIKE if the only bikes available are shared bikes.

Xuan-1998 committed:
fix bugs

Xuan-1998 committed:
testing on the changes

Xuan-1998 committed:
only not reaturning BIKE when all are just bikes

Xuan-1998 committed:
creating available vehicles filter

Xuan-1998 committed:
creating helper for isShared

Xuan-1998 committed:
create available vehicle filter

Xuan-1998 committed:
merge develop and update gradle

Xuan-1998 committed:
addressing peer advice

Xuan-1998 committed:
fix compilation

Xuan-1998 committed:
fix compilation

Xuan-1998 committed:
deleting useless

zneedell committed:
Merge pull request #3600 from LBNL-UCB-STI/xj/zn/tour-mode-choice-strategy

Xuan-1998 commented:
Let's merge to Dima's branch, and if Dima could do a last round check it would be wonderful
